openshift_metrics/utils.py

```diff
@@ -262,26 +291,28 @@ def condense_metrics(input_metrics_dict, metrics_to_check):
 
         start_epoch_time = epoch_times_list[0]
 
-        # calculate the interval if we have more than 1 measurement, otherwise
-        # use the STEP_MIN from the query as best guess
-        if len(epoch_times_list) > 1:
-            interval = epoch_times_list[1] - epoch_times_list[0]
```

Comment:
I'm a bit confused by this logic. In the previous version, we calculate the distance between readings by subtracting the timestamps. Assuming we do the readings at the same time, that seems to be a good approximation of the interval and if a pod disappears before the next reading, it seems reasonable to assume that it was running for `<= last_reading_time + time_between_last_readings`.

In the new logic, we seem to hardcode the step between readings to 30 minutes and disregard the actual time between readings, adding a hardcoded interval to everything.

Don't we still want to do that computation?

Reply:
The difference between the 2 timestamps is how I determine that a reading was skipped to ultimately determine that a pod was not running during that time.

I am hard-coding the interval to STEP_MIN * 60 seconds (15 mintutes * 60 = 900 seconds) where `STEP_MIN` is the frequency at which we collect metrics from prometheus. So, the time-series data in prometheus timestamps is always 15 minutes apart when the pod is running. Now, if any 2 consecutive timestamps are more than 15 minutes apart then that means there was no metric collected which means the pod was not running that time.

I had added that logic when I was testing different frequencies of collecting prometheus metrics, but I have settled on 15 minutes and have been using that for the metrics we've collected in the last 6 months; so this would still be accurate if we ran it against those metrics.

> In the new logic, we seem to hardcode the step between readings to 30 minutes and disregard the actual time between readings, adding a hardcoded interval to everything.

The actual times aren't disregarded, we still calculate the duration by subtracting the epoch times at the start and stop times for a pod, and then adding the interval makes it just a bit more accurate. If we didn't add the interval then for a pod that run for exactly 1 day it'll report it as 23 hours and 45 minutes instead of 24 hours.

Reply:
If the timestamps are always 15 minutes apart and there is no point in recomputing interval since it's unchanging (except for checking if the distance is longer than interval), I'm fine with removing that piece of logic. My comment did not assume that the intervals would always be the same value.

Would there ever be the case that a reading is not done exactly 15 minutes apart? Say 15 minutes and 1 second?

Reply:
> Would there ever be the case that a reading is not done exactly 15 minutes apart? Say 15 minutes and 1 second?

Nope, that's just how prometheus is returning the data to me. The time-series is always in 15 minutes increment (or whatever interval you ask it).

Reply:
Sounds good then.